### DISCUSSION TITLE
운영체제 - 프로세스 동기화 & 데드락

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#146] [질의 응답]
질문&답변 달기~

**Comment by watty-exclamationMark** [comment]
## Q. 프로세스 동기화란 무엇인가요?
다중 프로세스 환경에서 여러 개의 프로세스가 공유 자원에 접근하는 것을 조정하는 것입니다.

**Reply by sikutisa** [reply]
## Q. 프로세스 동기화가 왜 필요한가요?
여러 이유가 있겠지만, 다중 프로세스 환경에서는 race condition이 발생할 수 있습니다. 프로세스가 공동 데이터에 접근하는 순서에 따라 전혀 의도치 않은 결과가 나올 수 있습니다.

**Reply by seongeun42** [reply]
## Q. Race Condition이 무엇인가요?
여러 프로세스 혹은 스레드가 공유자원에 동시에 접근할 때, 공유자원에 대한 접근 순서에 따라 예상했던 실행 결과가 달라질 수 있는 상황을 의미합니다.

예를 들어 A와 B가 값이 3인 count라는 변수에 동시에 접근해 값을 +1을 해 변경하게 되면, 기대한 결과는 5인데 A와 B 모두 값이 3인 상황에서 동시 접근해 변경했기 때문에 한 결과가 덮어씌워져 4가 되게 됩니다.

**Reply by seongeun42** [reply]
## Q. 공유자원이 무엇인가요?

공유자원이란 여러 프로세스 혹은 스레드가 공동으로 이용하는 변수, 메모리, 파일 등을 말합니다.

**Reply by seongeun42** [reply]
## Q. 임계구역이란 무엇인가요?

공유 자원 접근 순서에 따라 실행 결과가 달라질 수 있는 프로그램의 영역을 임계 구역이라고 합니다.

**Comment by sikutisa** [comment]
## Q. 데드락이 무엇인가요?
두 개 이상의 작업이 서로 상대방의 작업이 끝나기 만을 기다리고 있기 때문에 결과적으로 아무것도 완료되지 못하는 상태를 의미합니다.

**Reply by watty-exclamationMark** [reply]
## Q. 데드락의 발생 조건이 무엇인가요?
상호 대기, 점유 대기, 비선점, 순환 대기 이 네 가지의 상태 모두 만족하지 않으면 데드락은 발생하지 않습니다.

**Comment by zini9188** [comment]
## Q. 세마포어(Semaphore)란 무엇인가요?
세마포어는 공유 자원에 동시에 접근할 수 있는 스레드/프로세스의 수를 변수로 제한하는 상호 배제를 위한 동기화 기법입니다.

**Comment by seongeun42** [comment]
## Q. 뮤텍스란 무엇인가요?
뮤텍스란 임계구역에 한 번에 하나의 스레드 혹은 프로세스만 접근할 수 있도록 제한하는 동기화 기법 중 하나입니다.

**Comment by Jihn0118** [comment]
## Q. 교착 상태의 해결 방법에는 어떤 것들이 있나요?
교착 상태 예방, 교착 상태 회피, 교착 상태 탐지, 교착 상태 회복이 있습니다.
